src/test/rpc/KeyGeneration_test.cpp

```diff
@@ -388,8 +386,8 @@ class WalletPropose_test : public ripple::TestSuite
 
             auto ret = keypairForSignature(params, error);
             BEAST_EXPECT(!contains_error(error));
-            BEAST_EXPECT(ret.first.size() != 0);
-            BEAST_EXPECT(ret.first == publicKey);
+            BEAST_EXPECT(ret->first.size() != 0);
+            BEAST_EXPECT(ret->first == publicKey);
```

Comment:
I think `BEAST_EXPECT(ret);` is a better way to communicate that `ret` must not be unseated. What do you think?

Reply:
Better than what? My suggestion includes `BEAST_EXPECT(ret)`, and does communicate that it must be seated. If you're saying it would be better as a standalone function call, then I disagree. If `ret` is somehow unseated, then having the `if` statement will still cause the test to fail, but the suite will continue. If it's standalone, and `ret` is somehow unseated, then the next test, which dereferences `ret`, will `assert`, crash, or have UB.

Reply:
But all the `ret`'s are in separate scopes. Even if one of the `ret` is unseated, I don't believe it crashes the subsequent tests.

Depending on the nature of the test case, only `if` or the `else` portion of the code is relevant. For example, this [test](https://github.com/ckeshava/rippled/blob/7df60ffd858d0f0f65aca8d5f5ff972887bf4f2c/src/test/rpc/KeyGeneration_test.cpp#L708) must necessarily have an unseated `ret`.

This [test](https://github.com/ckeshava/rippled/blob/7df60ffd858d0f0f65aca8d5f5ff972887bf4f2c/src/test/rpc/KeyGeneration_test.cpp#L389) must have a valid value inside `ret` for correctness. What should I write in the else portion of these tests?

Reply:
> But all the `ret`'s are in separate scopes. Even if one of the `ret` is unseated, I don't believe it crashes the subsequent tests.

It asserts in Windows builds. That counts as a crash.

https://en.cppreference.com/w/cpp/utility/optional/operator*
"The behavior is undefined if `*this` does not contain a value. "

Potential UB is unacceptable except in the rarest of circumstances, and this isn't one of them.

> This [test](https://github.com/ckeshava/rippled/blob/7df60ffd858d0f0f65aca8d5f5ff972887bf4f2c/src/test/rpc/KeyGeneration_test.cpp#L389) must have a valid value inside `ret` for correctness. What should I write in the else portion of these tests?

Nothing. If the `ret` is unseated, there's nothing else to check. If you use the `if(BEAST_EXPECT(ret))` construct, the suite will be marked as failed, and you'll get that result at the end of the run.

Reply:
ok, this makes it clear, thanks. I'll update the code

Reply:
hello Ed, I have addressed these comments in f69b0d0bdf62d6aa7cfb182decbc3ded350a3422

Reply:
I believe the negative tests do not need any other modification (`BEAST_EXPECT(!ret);` cases) because there is no opportunity for undefined behavior.